### PR TITLE
D8CORE-105: Local footer.

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Creates a responsive local footer component
- Adds FA assets 
- Adds a link list macro
- Adds signup form component
- Updates Dependencies
- Adds tugboat

# Review By (Date)
- Please review by October 18

# Urgency
- Med-high as we would like to get it deployed to our pilot at the end of the month

# Steps to Test

1. Check out this branch
2. Update npm dependencies `npm install`
3. Compile code `npm run build`
4. Compile styleguide `npm run styleguide`
5. Review code
6. Compare to mock: https://www.figma.com/file/Kmd4utmJFPRMVeCFEEBQhLtx/Stanford-Design-Library?node-id=5268%3A11637
7. Reivew for accessibility

# Affected Projects or Products
- Decanter
- D8 Core project

# Associated Issues and/or People
- https://github.com/SU-SWS/decanter/issues/491
- @kerri-augenstein 

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by yvonnetangsu** [comment]
FYI @sherakama , JB said we probably want to check with Dana Granoski, our brand designer on the local footer design first.

**Comment by sherakama** [comment]
@kerri-augenstein Heads up.

**Comment by sherakama** [comment]
@yvonnetangsu On a slightly different tangent, I figured out a way to get FA into this PR without adding ALL of the markup. I did have to add the fonts to the default decanter.css assets but nothing shows up in the decanter-no-markup which is good as we have access to their sass mixins/vars/etc.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu On a slightly different tangent, I figured out a way to get FA into this PR without adding ALL of the markup. I did have to add the fonts to the default decanter.css assets but nothing shows up in the decanter-no-markup which is good as we have access to their sass mixins/vars/etc.

Sounds great - thanks, @sherakama !

**Comment by yvonnetangsu** [comment]
Quick comments - 

1. You're so fast! 😄 
2. Perhaps add a gap between the two navs (in the middle cell) that's the same width as the grid gap. So if some links get long, they don't run into each other.
3. For prefooter, we keep the font weight on links 400 (Regular) except for the action links - though we could wait to change these things till Kerri finalize things with Dana.
4. I have some code written for the social media icons that I could share (JB wants them to show brand color when hovered over).

I'll do a more in-depth review soon.

**Comment by yvonnetangsu** [comment]
And it looks like the action link in the local footer has a different icon than the one in the link component section? Should we add that as a link variant?

**Comment by yvonnetangsu** [comment]
```
// Social media brand colors
$social-brands: (
    'Facebook':                    #3b579d,
    'Twitter':                     #1da1f2,
    'Instagram':                   #d73676,
    'LinkedIn':                    #0077b5,
    'YouTube':                     #cd201f,
    'iTunes U':                    #a3aaae,
    'Reddit':                      #ff5700,
);
```
I copied the above from our WP theme's color map - please feel free to use them for the local footer. We use them as hover/active colors. Icon normal states are black.

**Comment by yvonnetangsu** [comment]
@sherakama , seems like the headings "links to", "resources for", "Sign up for our email" could be `h2` instead of `<p>` so the voice over rotor can list them as headings? 

@kerri-augenstein, in the Figma design, "Sign up for our email" has larger font size compared than "links to" and "resources for". Is it possible to make them the same (maybe the larger size)?

I talked to JB and we're modifying the WP theme to use the local footer design. However, using our WP template, site builders can put in custom content and widgets, and they each can have a header - so I'm wondering what's the official "header" style for the local footer.

**Comment by kerri-augenstein** [comment]
@yvonnetangsu @sherakama yep, I hear ya on the headers. 

Let's modify all of them to be 18px. No need for any changes on any breakpoints.

**Comment by sherakama** [comment]
> And it looks like the action link in the local footer has a different icon than the one in the link component section? Should we add that as a link variant?

I am using the link-icon mixin for this element.

**Comment by yvonnetangsu** [comment]
Quick question about the lockup in the local footer - is it only going to display line 1 regardless of how many lines the full lockup in the header is for the department?

**Comment by kerri-augenstein** [comment]
@yvonnetangsu great question! To keep it simple, we would mirror whatever lockup is present in the header. So it would include any text/styles/lines. But. There is no reason you would/should need to include this in your Wordpress platform if your sites don't require the lockup.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu great question! To keep it simple, we would mirror whatever lockup is present in the header. So it would include any text/styles/lines. But. There is no reason you would/should need to include this in your Wordpress platform if your sites don't require the lockup.

Thanks, @kerri-augenstein! I'll follow your design as closely as possible in our WP theme and will include the lockup if there is one.

**Comment by sherakama** [comment]
> Quick question about the lockup in the local footer - is it only going to display line 1 regardless of how many lines the full lockup in the header is for the department?

The default is to use option-a but everything is a variable so you could pass in any of the lockup options.

**Comment by sherakama** [comment]
@yvonnetangsu @kerri-augenstein 

Feedback changes are up. Please have another pass.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu @kerri-augenstein
> 
> Feedback changes are up. Please have another pass.

I'm taking the afternoon off after the interview today. Will look at this on Monday. Thanks for the changes!

**Comment by kerri-augenstein** [comment]
@sherakama if I pull down the latest, is there a spot I can go to look at this?

**Comment by yvonnetangsu** [comment]
I just remember I have a question for @kerri-augenstein - is there a reason the two link blocks are stacked vertically for the MD break point and not side by side so there's less empty space on the right below the email signup? I'm looking at the Figma design.

**Comment by kerri-augenstein** [comment]
@yvonnetangsu oh, I hadn't thought of that. Here's a version to think about: 
https://www.figma.com/file/Kmd4utmJFPRMVeCFEEBQhLtx/Stanford-Design-Library?node-id=5524%3A212

And this actually makes me think about how all of the variable blocks will collapse depending on what goes into the columns. 

@sherakama which makes me wonder about that idea you've had for a long time, about giving the authors control over how the responsive design is implemented across all breakpoints. ...

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu oh, I hadn't thought of that. Here's a version to think about:
> https://www.figma.com/file/Kmd4utmJFPRMVeCFEEBQhLtx/Stanford-Design-Library?node-id=5524%3A212
> 
> And this actually makes me think about how all of the variable blocks will collapse depending on what goes into the columns.
> 
> @sherakama which makes me wonder about that idea you've had for a long time, about giving the authors control over how the responsive design is implemented across all breakpoints. ...

We will likely have our own custom responsive behavior for the local footer for the WP theme, since it's a template that anyone can use, so they can put a 100 widgets in the footer if they choose to do so 😄 Also, no blocks in the footer is required - they don't need to add contact or social media if they don't want to.

**Comment by yvonnetangsu** [comment]
@sherakama , I haven't looked at this so perhaps you have it already - is it possible to provide an option for the whole local footer content area (where the blocks go) an open HTML field?

**Comment by sherakama** [comment]
> @sherakama , I haven't looked at this so perhaps you have it already - is it possible to provide an option for the whole local footer content area (where the blocks go) an open HTML field?

The `<header>` is completely optional with an open variable field for arbitrary content
The 3 cells have two options for custom manipulation; An open variable for arbitrary content and a twig block region. The cells and their responsive behavior are not as flexible due to the way the elements need to be re-arranged in this responsive behavior. The content is flexible but the scaffolding is not. I left a plethora of comments in the twig code for clarity.

**Comment by yvonnetangsu** [comment]
> > @sherakama , I haven't looked at this so perhaps you have it already - is it possible to provide an option for the whole local footer content area (where the blocks go) an open HTML field?
> 
> The `<header>` is completely optional with an open variable field for arbitrary content
> The 3 cells have two options for custom manipulation; An open variable for arbitrary content and a twig block region. The cells and their responsive behavior are not as flexible due to the way the elements need to be re-arranged in this responsive behavior. The content is flexible but the scaffolding is not. I left a plethora of comments in the twig code for clarity.

We might need an option for the whole content area inside `<section class="su-local-footer__columns"></section>` to be an open markup field, since we might not be able to adopt the same responsive behavior of the cells (we can't limit how many cells/widgets people can put into the footer area in the WP template). No need to write styles for the open field since we'll make that specific to our WP template. So basically we'll use `<section class="su-local-footer__columns"></section>` but we don't need the `<div class="su-local-footer__cell1">` etc, instead we'll have WP inject any widgets that are added as individual `<div>`. Is that possible?

**Comment by sherakama** [comment]
There comes a point where you are not really getting much benefit from the component. If you are not planning on using the responsive layout or the contents of the cells you aren't really using the component and might just be adding extra complication to your build. Most of the css and html markup is for the contents. I could pull the signup form into its own component if you plan to re-use that part.

**Comment by yvonnetangsu** [comment]
> There comes a point where you are not really getting much benefit from the component. If you are not planning on using the responsive layout or the contents of the cells you aren't really using the component and might just be adding extra complication to your build. Most of the css and html markup is for the contents. I could pull the signup form into its own component if you plan to re-use that part.

You got a point . I wanted to use the twig file just so I feel like I'm getting more used out of Decanter 😄 Ok, let's leave it as is then. I don't think I need the signup form for now.
 I'm out sick today so I'll review this tomorrow. I'll still send the feedback to John by EOD.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu oh, I hadn't thought of that. Here's a version to think about:
> https://www.figma.com/file/Kmd4utmJFPRMVeCFEEBQhLtx/Stanford-Design-Library?node-id=5524%3A212

@kerri-augenstein , I was actually thinking more about the layout at the MD breakpoint. Could the two link lists be side by side since the left column is a lot longer?

Screen Shot 2019-10-15 at 9 48 48 AM

**Comment by sherakama** [comment]
@yvonnetangsu pending the change to the MD breakpoint list columns is the rest of the code looking good?

**Comment by sherakama** [comment]
> I'll review this for real today, and I believe the links are going to be bright red (I think Dana is ok with the links being red, but have concerns more about the red buttons).

Super, thanks. I left the links the default color and I am totally happy to leave them as blue in Decanter. Let me know on the decision. either way.

File: core/src/scss/components/local-footer/_local-footer.scss [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  }
+
+  // Font Awesome icons for specific selectors.
+  // ------------------------------------------------
+  .su-local-footer__social-facebook {
+    &::before {
+      content: fa-content($fa-var-facebook-f);
+    }
+
+    &:hover {
```
Comment (yvonnetangsu) [inline core/src/scss/components/local-footer/_local-footer.scss]:
```suggestion
    &:hover,
    &:focus {
```

Comment (yvonnetangsu) [reply core/src/scss/components/local-footer/_local-footer.scss]: Could we add the color change for on focus also (and for all the other social icons please)?

File: core/src/templates/components/local-footer/local-footer.twig [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      {% block contentcell3 %}
+      {#
+        The signup form is optional. If it doesn't meet your content needs
+        use the cell3 variable instead.
+      #}
+      {% if signup_form_action is not empty %}
+      {# Add any additional markup above the form with this variable. #}
+      {{ signup_form_content }}
+      <form class="su-local-footer__form" action="{{ signup_form_action }}" method="{{ signup_form_method|default('post') }}">
+        <input type="email" id="{{ signup_form_field_email }}" name="{{ signup_form_field_email }}" placeholder="{{ signup_form_field_email_placeholder|default('email address') }}" {{ signup_form_field_email_required|default('required') }} />
```
Comment (yvonnetangsu) [inline core/src/templates/components/local-footer/local-footer.twig]: Could we add a screen reader-only label for this input field? It's an a11y thing. We did that for the site search component also.

**Comment by sherakama** [comment]
> I believe for the h2 headers in the local footer, @kerri-augenstein mentioned that she decided to go with 18px at all breakpoints. Also, the "normal" font size in her footer is smaller than the body font size in her Figma mockup. In our WP template, I made base font size for sidebar and footer just a little bit smaller than the body for all the breakpoint ranges.

Is there a particular reason why we wouldn't want to use one of our standardized and systemized heading styles? One or two pixel sizes may be significant enough to warrant a deviation from the standard set of heading types we have created but I would like to hear why as it seems like my place to ask these kinds of things.

**Comment by yvonnetangsu** [comment]
> Is there a particular reason why we wouldn't want to use one of our standardized and systemized heading styles? One or two pixel sizes may be significant enough to warrant a deviation from the standard set of heading types we have created but I would like to hear why as it seems like my place to ask these kinds of things.

Let's discuss this with @kerri-augenstein . Personally, I'm ok with either @modular-typography(1) (or equivalent @type-?) or base font size @modular-typography(0) even.

**Comment by yvonnetangsu** [comment]
@sherakama , new changes look good! Thanks!

**Comment by sherakama** [comment]
@kerri-augenstein three questions for you:

1. Why the rounded login button vs square signup button?
2. Layout at the MD breakpoint, should the link lists be side x side?
3. The h2 headers are 18px across all breakpoints, can they be one of the modular type scales instead?

**Comment by yvonnetangsu** [comment]
@sherakama , I just tested with IE11 and there are some layout issues.

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-ykuol77mxelou2fp6bavkcglbzbhawk5.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5da9faec44c5fa0dedf0913c

**Comment by kerri-augenstein** [comment]
hey y'all, @sherakama @yvonnetangsu here are my thoughts on the questions ya dropped 

1. Why the rounded login button vs square signup button?
Layout at the MD breakpoint, should the link lists be side x side?
The h2 headers are 18px across all breakpoints, can they be one of the modular type scales instead?

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-ykuol77mxelou2fp6bavkcglbzbhawk5.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5da9faec44c5fa0dedf0913c

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-pnfebndundjdki2go65af50tyltcbv3w.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
@kerri-augenstein @yvonnetangsu 

1. I left the login button as rounded. 
2. I put the two lists of links as side x side on the MD breakpoint
3. Headers are all 18px now for all breakpoints

I have also abstracted the signup-form into its own component as Kerri has identified projects where it will be used in multiple places. I'll work on the moving placeholder text for the form next.

**Comment by sherakama** [comment]
Looks like moving placeholders will take a bit of work. I'm going to create a ticket for it so that it doesn't block this PR.

**Comment by sherakama** [comment]
https://github.com/SU-SWS/decanter/issues/519

**Comment by yvonnetangsu** [comment]
Thanks for doing this, @sherakama . I'll find some time to find Kevin's code for the signup form label soon. Should I push to this branch directly?

**Comment by yvonnetangsu** [comment]
@sherakama sorry, please ignore above comment. I misunderstood placeholder as sass placeholders 😄 . Shall I push the placeholder/label code to the new branch once you set it up?

**Comment by sherakama** [comment]
I would prefer to have this merged sooner than later so if I could pick I would ask that you spend your time on review first and then we get to the moving placeholders. I'm happy to move this work to the v6 branch if you don't feel that it meets our accessibility standards. IMHO, the moving placeholder text is an improvement and not an accessibility blocker.

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-xpopafvmlobtngyqmfwlda1qf4itzmhe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daa54b944c5fabd83f26527

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-muh4cxmr6boknwuz7jy0tth1ubpvivps.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5daf976344c5fa301a0886bf

Visual Diffs:
- https://dashboard.tugboat.qa/5daf976344c5fa301a0886bf/visualdiffs

**Comment by yvonnetangsu** [comment]
@sherakama , this is looking really great - about perfect on IE except that white space that doesn't take on the fog color background on the right, but I'm fine with it since it's IE. Responsive behavior, heading sizes look great. There are some weird issues with Edge 18 but looks like `.kss-fullscreen-mode` need a width:100%; and that would solve the problem.

I'm thinking that some of the remaining design related tweaks (button styles, link colors) can be worked on in another PR after we hear Dana/Kerri's final decisions.

That leaves the only thing that's on my mind - currently the normal font size for the local footer just follows the body font size (19px on desktop), but I believe Kerri has it a bit smaller for the local footer. I'm also fine with finetuning this in another PR if you're eager to merge this.

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-eyysoeycoey79ujuejoihz6lpm4nxowj.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5db087aa44c5fa544a0de0d6

Visual Diffs:
- https://dashboard.tugboat.qa/5db087aa44c5fa544a0de0d6/visualdiffs

**Comment by sherakama** [comment]
@yvonnetangsu 

Ok, I made a few font-size related tweaks, but if we could say this is good enough, I'll change the base branch to v6 and we can merge it in to there as we wait for the feedback on buttons and link colors.

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-eyysoeycoey79ujuejoihz6lpm4nxowj.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5db087aa44c5fa544a0de0d6

Visual Diffs:
- https://dashboard.tugboat.qa/5db087aa44c5fa544a0de0d6/visualdiffs

**Comment by sherakama** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr510-eyysoeycoey79ujuejoihz6lpm4nxowj.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5db087aa44c5fa544a0de0d6

Visual Diffs:
- https://dashboard.tugboat.qa/5db087aa44c5fa544a0de0d6/visualdiffs

**Comment by sherakama** [comment]
Thank you so much for all of the review!

**Pull request merged**